controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
+	cache *rc.ObjectCache,
+	client client.Client,
+	recorder record.EventRecorder,
+	app *crd.ClowdApp,
+	log logr.Logger,
+	req *ctrl.Request) ClowdAppReconciliation {
+	return ClowdAppReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      &ctx,
+		client:   client,
+		log:      &log,
+		app:      app,
+		req:      req,
+	}
+}
+
+func (r *ClowdAppReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.getApp,
+		r.setPresentAndManagedApps,
+		r.startMetrics,
+		r.isAppMarkedForDeletion,
+		r.addFinalizer,
+		r.isEnvLocked,
+		r.isAppDisabled,
+		r.getClowdEnv,
+		r.isEnvCoreNamespaceDeleted,
+		r.isAppCoreNamespaceDeleted,
+		r.isClowdEnvReconciled,
+		r.isEnvReady,
+		r.runProviders,
+		r.applyCache,
+		r.setAppResourceStatus,
+		r.deletedUnusedResources,
+		r.setReconciliationSuccessful,
+		r.stopMetrics,
+	}
+}
+
+func (r *ClowdAppReconciliation) Reconcile() (ctrl.Result, error) {
+	for _, step := range r.steps() {
+		result, err := step()
+		if err != nil {
+			return result, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) startMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics = ReconciliationMetrics{}
+	r.reconciliationMetrics.init(r.app.Name, r.app.Spec.EnvName)
+	r.reconciliationMetrics.start()
+
+	if clowderconfig.LoadedConfig.Features.PerProviderMetrics {
+		requestMetrics.With(prometheus.Labels{"type": "app", "name": r.app.GetIdent()}).Inc()
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) stopMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics.stop()
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) setPresentAndManagedApps() (ctrl.Result, error) {
+	if _, ok := presentApps[r.app.GetIdent()]; !ok {
+		presentApps[r.app.GetIdent()] = true
+	}
+	presentAppsMetric.Set(float64(len(presentApps)))
+
+	delete(managedApps, r.app.GetIdent())
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) getApp() (ctrl.Result, error) {
+	if getAppErr := r.client.Get(*r.ctx, r.req.NamespacedName, r.app); getAppErr != nil {
+		if k8serr.IsNotFound(getAppErr) {
+			// Must have been deleted
+			return ctrl.Result{}, errors.New(SKIPRECONCILE)
+		}
+		r.log.Info("App not found", "env", r.app.Spec.EnvName, "app", r.app.GetIdent(), "err", getAppErr)
+		return ctrl.Result{}, getAppErr
+	}
+	//This is kinda side-effecty but I couldn't think of a better place
+	//to put it.
+	logWithEnv := r.log.WithValues("env", r.app.Spec.EnvName)
+	r.log = &logWithEnv
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isAppMarkedForDeletion() (ctrl.Result, error) {
+	isAppMarkedForDeletion := r.app.GetDeletionTimestamp() != nil
```

Comment:
As soon as we get this in, I want to migrate this out and into the rhc-osdk-utils - No action, just an FYI

Reply:
Good call :+1: